File: Decyx/decyx/dialogs/caller_selection_dialog.py
# caller_selection_dialog.py

import threading
import javax.swing as swing
from javax.swing import JFrame, JPanel, JCheckBox, JButton, JScrollPane, BoxLayout, JLabel, JTextArea
from java.awt import BorderLayout, Dimension

from decyx.decompiler import decompile_function

# Dialog Class
class CallerSelectionDialog(JFrame):
    """
    Dialog for selecting which caller functions' code to include for additional context.
    """
    def __init__(self, callers, current_program, monitor, on_selection_complete):
        super(CallerSelectionDialog, self).__init__("Select Callers to Include")
        self.callers = callers
        self.current_program = current_program
        self.monitor = monitor
        self.on_selection_complete = on_selection_complete
        self.selected_callers = []
        self.init_ui()

    def init_ui(self):
        try:
            panel = JPanel(BorderLayout())

            self.add_callers_panel(panel)
            self.add_preview_panel(panel)
            self.add_buttons_panel(panel)

            self.getContentPane().add(panel)
            self.setSize(800, 400)
            self.setLocationRelativeTo(None)
            self.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE)
            self.setVisible(True)
        except Exception as e:
            print "Error initializing CallerSelectionDialog UI: {}".format(e)

    def add_callers_panel(self, panel):
        try:
            callers_panel = JPanel()
            callers_panel.setLayout(BoxLayout(callers_panel, BoxLayout.Y_AXIS))
            self.caller_checkboxes = []
            for caller in self.callers:
                checkbox = JCheckBox(caller.getName())
                checkbox.setSelected(False)
                checkbox.setToolTipText("Include code from caller: {}".format(caller.getName()))
                # Bind the caller and checkbox to the action listener using lambda with default arguments
                checkbox.addActionListener(lambda e, c=caller: self.update_preview(c, e))
                self.caller_checkboxes.append(checkbox)
                callers_panel.add(checkbox)

            scroll_callers = JScrollPane(callers_panel)
            scroll_callers.setPreferredSize(Dimension(200, 300))
            panel.add(scroll_callers, BorderLayout.WEST)
        except Exception as e:
            print "Error adding callers panel: {}".format(e)

    def update_preview(self, caller, event):
        try:
            if event.getSource().isSelected():
                decompiled_code, _ = decompile_function(caller, self.current_program, self.monitor)
                if decompiled_code:
                    code_length = len(decompiled_code)
                    self.update_preview_label(code_length)
                    self.set_preview_text(decompiled_code)
                else:
                    self.set_preview_text("Decompilation failed or no code available.")
                    self.update_preview_label(0)
            else:
                self.set_preview_text("Select a caller to see its decompiled code preview.")
                self.update_preview_label(0)
        except Exception as e:
            print "Error updating preview: {}".format(e)

    def add_preview_panel(self, panel):
        try:
            preview_panel = JPanel(BorderLayout())
            self.preview_label = JLabel("Caller Function Preview:")
            preview_panel.add(self.preview_label, BorderLayout.NORTH)

            self.preview_text_area = JTextArea()
            self.preview_text_area.setEditable(False)
            self.preview_text_area.setLineWrap(True)
            self.preview_text_area.setWrapStyleWord(True)
            self.preview_text_area.setText("Select a caller to see its decompiled code preview.")
            self.preview_text_area.setToolTipText("Decompiled code preview of the selected caller")
            self.preview_text_area.setCaretPosition(0)
            scroll_preview = JScrollPane(self.preview_text_area)
            preview_panel.add(scroll_preview, BorderLayout.CENTER)

            panel.add(preview_panel, BorderLayout.CENTER)
        except Exception as e:
            print "Error adding preview panel: {}".format(e)

    def add_buttons_panel(self, panel):
        try:
            buttons_panel = JPanel()
            buttons_panel.setLayout(BoxLayout(buttons_panel, BoxLayout.Y_AXIS))

            select_buttons = JPanel()
            select_all_button = JButton("Select All Callers")
            select_all_button.addActionListener(lambda e: self.select_all(True))
            select_all_button.setToolTipText("Select all caller checkboxes")
            unselect_all_button = JButton("Unselect All Callers")
            unselect_all_button.addActionListener(lambda e: self.select_all(False))
            unselect_all_button.setToolTipText("Unselect all caller checkboxes")
            select_buttons.add(select_all_button)
            select_buttons.add(unselect_all_button)

            action_buttons = JPanel()
            ok_button = JButton("OK")
            ok_button.addActionListener(lambda e: self.ok())
            ok_button.setToolTipText("Confirm selection and proceed")
            cancel_button = JButton("Cancel")
            cancel_button.addActionListener(lambda e: self.cancel())
            cancel_button.setToolTipText("Cancel selection")
            action_buttons.add(ok_button)
            action_buttons.add(cancel_button)

            buttons_panel.add(select_buttons)
            buttons_panel.add(action_buttons)
            panel.add(buttons_panel, BorderLayout.SOUTH)
        except Exception as e:
            print "Error adding buttons panel: {}".format(e)

    def set_preview_text(self, text):
        try:
            self.preview_text_area.setText(text)
            self.preview_text_area.setCaretPosition(0)
        except Exception as e:
            print "Error setting preview text: {}".format(e) 

    def update_preview_label(self, length):
        try:
            self.preview_label.setText("Caller Function Preview (Length: {} characters):".format(length))
        except Exception as e:
            print "Error updating preview label: {}".format(e)

    def select_all(self, select):
        try:
            for checkbox in self.caller_checkboxes:
                checkbox.setSelected(select)
        except Exception as e:
            print "Error selecting all callers: {}".format(e)

    def ok(self):
        try:
            self.selected_callers = [
                caller for checkbox, caller in zip(self.caller_checkboxes, self.callers) 
                if checkbox.isSelected()
            ]
            self.on_selection_complete(self.selected_callers)
            self.dispose()
        except Exception as e:
            print "Error in OK action: {}".format(e)

    def cancel(self):
        try:
            self.selected_callers = []
            self.on_selection_complete(self.selected_callers)
            self.dispose()
        except Exception as e:
            print "Error in Cancel action: {}".format(e)

# Helper Functions
def show_caller_selection_dialog(callers, current_program, monitor):
    """
    Displays the CallerSelectionDialog and waits for user interaction.

    Args:
        callers (list): List of caller Function objects.
        current_program (Program): The current program context.
        monitor (TaskMonitor): The monitor object for progress tracking.

    Returns:
        list: The list of selected caller Function objects.
    """
    selected_callers = []
    dialog_complete = threading.Event()

    def on_selection(selected):
        selected_callers.extend(selected)
        dialog_complete.set()

    def create_dialog():
        CallerSelectionDialog(callers, current_program, monitor, on_selection)

    swing.SwingUtilities.invokeLater(create_dialog)
    dialog_complete.wait()
    return selected_callers

File: Decyx/decyx/dialogs/CallerSelectionDialog.java
package decyx.dialogs;

import decyx.DecompilerHelper;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.Program;
import ghidra.util.task.TaskMonitor;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;

/**
 * Dialog for selecting which caller functions' code to include for additional context.
 */
public class CallerSelectionDialog extends JFrame {

    private final List<Function> callers;
    private final Program currentProgram;
    private final TaskMonitor monitor;
    private final Runnable onComplete;
    private final List<Function> selectedCallers = new ArrayList<>();
    private final List<JCheckBox> callerCheckboxes = new ArrayList<>();
    private JTextArea previewTextArea;
    private JLabel previewLabel;

    public CallerSelectionDialog(List<Function> callers, Program currentProgram,
                                  TaskMonitor monitor, Runnable onComplete) {
        super("Select Callers to Include");
        this.callers = callers;
        this.currentProgram = currentProgram;
        this.monitor = monitor;
        this.onComplete = onComplete;
        initUI();
    }

    private void initUI() {
        try {
            JPanel panel = new JPanel(new BorderLayout());

            addCallersPanel(panel);
            addPreviewPanel(panel);
            addButtonsPanel(panel);

            getContentPane().add(panel);
            setSize(800, 400);
            setLocationRelativeTo(null);
            setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
            setVisible(true);
        } catch (Exception e) {
            System.out.println("Error initializing CallerSelectionDialog UI: " + e.getMessage());
        }
    }

    private void addCallersPanel(JPanel panel) {
        JPanel callersPanel = new JPanel();
        callersPanel.setLayout(new BoxLayout(callersPanel, BoxLayout.Y_AXIS));

        for (Function caller : callers) {
            JCheckBox checkbox = new JCheckBox(caller.getName());
            checkbox.setSelected(false);
            checkbox.setToolTipText("Include code from caller: " + caller.getName());
            checkbox.addActionListener(e -> updatePreview(caller, checkbox.isSelected()));
            callerCheckboxes.add(checkbox);
            callersPanel.add(checkbox);
        }

        JScrollPane scrollCallers = new JScrollPane(callersPanel);
        scrollCallers.setPreferredSize(new Dimension(200, 300));
        panel.add(scrollCallers, BorderLayout.WEST);
    }

    private void updatePreview(Function caller, boolean selected) {
        try {
            if (selected) {
                DecompilerHelper.DecompileResult result =
                    DecompilerHelper.decompileFunction(caller, currentProgram, monitor, false);
                if (result != null && result.code != null) {
                    int codeLength = result.code.length();
                    updatePreviewLabel(codeLength);
                    setPreviewText(result.code);
                } else {
                    setPreviewText("Decompilation failed or no code available.");
                    updatePreviewLabel(0);
                }
            } else {
                setPreviewText("Select a caller to see its decompiled code preview.");
                updatePreviewLabel(0);
            }
        } catch (Exception e) {
            System.out.println("Error updating preview: " + e.getMessage());
        }
    }

    private void addPreviewPanel(JPanel panel) {
        JPanel previewPanel = new JPanel(new BorderLayout());
        previewLabel = new JLabel("Caller Function Preview:");
        previewPanel.add(previewLabel, BorderLayout.NORTH);

        previewTextArea = new JTextArea();
        previewTextArea.setEditable(false);
        previewTextArea.setLineWrap(true);
        previewTextArea.setWrapStyleWord(true);
        previewTextArea.setText("Select a caller to see its decompiled code preview.");
        previewTextArea.setToolTipText("Decompiled code preview of the selected caller");
        previewTextArea.setCaretPosition(0);
        JScrollPane scrollPreview = new JScrollPane(previewTextArea);
        previewPanel.add(scrollPreview, BorderLayout.CENTER);

        panel.add(previewPanel, BorderLayout.CENTER);
    }

    private void addButtonsPanel(JPanel panel) {
        JPanel buttonsPanel = new JPanel();
        buttonsPanel.setLayout(new BoxLayout(buttonsPanel, BoxLayout.Y_AXIS));

        JPanel selectButtons = new JPanel();
        JButton selectAllButton = new JButton("Select All Callers");
        selectAllButton.addActionListener(e -> selectAll(true));
        selectAllButton.setToolTipText("Select all caller checkboxes");
        JButton unselectAllButton = new JButton("Unselect All Callers");
        unselectAllButton.addActionListener(e -> selectAll(false));
        unselectAllButton.setToolTipText("Unselect all caller checkboxes");
        selectButtons.add(selectAllButton);
        selectButtons.add(unselectAllButton);

        JPanel actionButtons = new JPanel();
        JButton okButton = new JButton("OK");
        okButton.addActionListener(e -> ok());
        okButton.setToolTipText("Confirm selection and proceed");
        JButton cancelButton = new JButton("Cancel");
        cancelButton.addActionListener(e -> cancel());
        cancelButton.setToolTipText("Cancel selection");
        actionButtons.add(okButton);
        actionButtons.add(cancelButton);

        buttonsPanel.add(selectButtons);
        buttonsPanel.add(actionButtons);
        panel.add(buttonsPanel, BorderLayout.SOUTH);
    }

    private void setPreviewText(String text) {
        previewTextArea.setText(text);
        previewTextArea.setCaretPosition(0);
    }

    private void updatePreviewLabel(int length) {
        previewLabel.setText("Caller Function Preview (Length: " + length + " characters):");
    }

    private void selectAll(boolean select) {
        for (JCheckBox checkbox : callerCheckboxes) {
            checkbox.setSelected(select);
        }
    }

    private void ok() {
        selectedCallers.clear();
        for (int i = 0; i < callerCheckboxes.size(); i++) {
            if (callerCheckboxes.get(i).isSelected()) {
                selectedCallers.add(callers.get(i));
            }
        }
        onComplete.run();
        dispose();
    }

    private void cancel() {
        selectedCallers.clear();
        onComplete.run();
        dispose();
    }

    public List<Function> getSelectedCallers() {
        return selectedCallers;
    }

    /**
     * Displays the dialog and waits for user interaction.
     *
     * @param callers        list of caller Function objects
     * @param currentProgram the current program context
     * @param monitor        the monitor for progress tracking
     * @return the list of selected caller Function objects
     */
    public static List<Function> showDialog(List<Function> callers, Program currentProgram,
                                             TaskMonitor monitor) {
        CountDownLatch latch = new CountDownLatch(1);
        CallerSelectionDialog[] dialogHolder = new CallerSelectionDialog[1];

        SwingUtilities.invokeLater(() -> {
            dialogHolder[0] = new CallerSelectionDialog(callers, currentProgram, monitor, latch::countDown);
        });

        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        return dialogHolder[0] != null ? dialogHolder[0].getSelectedCallers() : new ArrayList<>();
    }
}
